src/runs.ts

```diff
@@ -0,0 +1,158 @@
+import { OpenAI } from "openai";
+import { RequestOptions } from "openai/core";
+import {
+  Run,
+  RunCreateParamsNonStreaming,
+  Runs,
+} from "openai/resources/beta/threads/runs/runs";
+import { LibrettoConfig, send_event } from ".";
+
+class RunObserver {
+  protected threadQueues: Record<string, string[]> = {};
+
+  constructor(protected client: OpenAI) {}
+
+  addRun(threadId: string, runId: string) {
+    const queue = this.threadQueues[threadId];
+
+    // If another run is already being processed for this thread, just enqueue this new run.
+    // The worker that's handling the existing run(s) will eventually process this one too.
+    if (queue && queue.length > 0) {
+      queue.push(runId);
+      return;
+    }
+
+    // No work is currently being done on this thread, so start a worker to process this run.
```

Comment:
by "start a worker" you mean to start an async function that just runs in the background?

Reply:
One thing I'm wondering is, if it is possible / likely that someone can createAndPoll a thread more than once in the same process? because it seems like then you'll have multiple background async functions each pulling items off of this.threadQueues[threadId] - because its synchronous I don't see any opportunities for race conditions or deadlocks the way it is written.. mostly you could just have 2 "workers" polling the same thread?

Might just be helpful to have some more comments explaining the flow